tests/functional/test_copy.py

```diff
@@ -0,0 +1,520 @@
+
+    def test_provide_copy_source_as_string(self):
+        self.copy_source = (
+            'mysourcebucket/mysourcekey?versionId=mysourceversionid'
+        )
+        expected_params = {
+            'Bucket': 'mysourcebucket',
+            'Key': 'mysourcekey',
+            'VersionId': 'mysourceversionid'
+        }
+
+        self.add_head_object_response(expected_params=expected_params)
+        self.add_successful_copy_responses()
+
+        future = self.manager.copy(**self.call_kwargs)
+        future.result()
+        self.stubber.assert_no_pending_responses()
+
+    def test_invalid_copy_source(self):
+        self.copy_source = ['bucket', 'key']
+        with self.assertRaises(TypeError):
+            self.manager.copy(**self.call_kwargs)
+
+    def test_provide_copy_source_client(self):
+        source_client = self.session.create_client(
+            's3', 'eu-central-1', aws_access_key_id='foo',
+            aws_secret_access_key='bar')
+        source_stubber = Stubber(source_client)
+        source_stubber.activate()
+        self.addCleanup(source_stubber.deactivate)
+
+        self.add_head_object_response(stubber=source_stubber)
+        self.add_successful_copy_responses()
+
+        call_kwargs = self.call_kwargs
+        call_kwargs['source_client'] = source_client
+        future = self.manager.copy(**call_kwargs)
+        future.result()
+
+        # Make sure that all of the responses were properly
+        # used for both clients.
+        source_stubber.assert_no_pending_responses()
+        self.stubber.assert_no_pending_responses()
+
+
+class TestNonMultipartCopy(BaseCopyTest):
+    __test__ = True
+
+    def test_copy(self):
+        expected_head_params = {
+            'Bucket': 'mysourcebucket',
+            'Key': 'mysourcekey'
+        }
+        expected_copy_object = {
+            'Bucket': self.bucket,
+            'Key': self.key,
+            'CopySource': self.copy_source
+        }
+        self.add_head_object_response(expected_params=expected_head_params)
+        self.add_successful_copy_responses(
+            expected_copy_params=expected_copy_object)
+
+        future = self.manager.copy(**self.call_kwargs)
+        future.result()
+        self.stubber.assert_no_pending_responses()
+
+    def test_copy_with_extra_args(self):
+        self.extra_args['MetadataDirective'] = 'REPLACE'
+
+        expected_head_params = {
+            'Bucket': 'mysourcebucket',
+            'Key': 'mysourcekey'
+        }
+        expected_copy_object = {
+            'Bucket': self.bucket,
+            'Key': self.key,
+            'CopySource': self.copy_source,
+            'MetadataDirective': 'REPLACE'
+        }
+
+        self.add_head_object_response(expected_params=expected_head_params)
+        self.add_successful_copy_responses(
+            expected_copy_params=expected_copy_object)
+
+        call_kwargs = self.call_kwargs
+        call_kwargs['extra_args'] = self.extra_args
```

Comment:
Hmm I will need to update this pattern in all of the tests then. It is a common pattern for upload, download, and copy. But I can do it.